### PR TITLE
Change input type to boolean numpy array

### PR DESCRIPTION
**Pull request by johnrudge**
At the moment, the `exact_cover` package requires the user to provide the input matrix as a matrix with `dtype=np.int32`. Exact cover problems are naturally described by boolean arrays, so I think it makes more sense to ask the user to supply an array with `dtype=bool`. This pull request changes the requested `dtype` to `bool`. This may break some codes which depend on exact_cover that are set up to give `dtype=np.int32` arrays, but those that use `dtype=DTYPE_FOR_ARRAY` should be fine. `dtype=bool` arrays also use a little less memory.

### CODE REVIEW

**Comment by jwg4** [comment]
Aren't bool arrays represented using 32-bit ints anyway?

This change still makes some sense because of ergonomics, but I think it would be far preferable to accept both array types, and perhaps Python list of lists too.

**Comment by johnrudge** [comment]
I agree, it would be good for the routine to happily to take both types of array. I've modified the wrapper to use numpy's `astype` to cast to the appropriate boolean type. I think numpy stores `bool` with 1 byte, and `int32` as 4 bytes.

File: README.md [hunk 1]
```diff
@@ -64,7 +64,7 @@ Now we can call `exact_cover`:
 
     >>> import numpy as np
     >>> import exact_cover as ec
-    >>> S = np.array([[1,0,0,1,0],[1,1,1,0,0],[0,1,1,0,0],[0,0,0,0,1]], dtype=np.int32)
+    >>> S = np.array([[1,0,0,1,0],[1,1,1,0,0],[0,1,1,0,0],[0,0,0,0,1]], dtype=bool)
```
Comment (jwg4) [inline README.md:67]: (we could import DTYPE_FOR_ARRAY here but for simplicity's sake this is probably better.)

File: exact_cover/helpers.py [hunk 2]
```diff
@@ -44,7 +44,7 @@ def create_numpy_array(a):
     >>> create_numpy_array([[0, 1], [1, 0]])
     numpy.array([[0, 1], [1, 0]])
     """
-    return np.array(a)
+    return np.array(a, dtype=DTYPE_FOR_ARRAY)
```
Comment (jwg4) [inline exact_cover/helpers.py:47]: 👍

File: exact_cover/wrapper.py [hunk 3]
```diff
@@ -1,10 +1,14 @@
 from exact_cover_impl import get_exact_cover as raw_get_exact_cover
 
 from .error import NoSolution
+from .io import DTYPE_FOR_ARRAY
 
 
 def get_exact_cover(matrix):
-    result = raw_get_exact_cover(matrix)
+    if matrix.dtype == DTYPE_FOR_ARRAY:
```
Comment (jwg4) [inline exact_cover/wrapper.py:8]: This is great. We should add a Python unit test of this functionality, eg the same simple exact cover problem with arrays of int32, bool, and explicit DTYPE_FOR_ARRAY.

File: tests/test_legacy_dlx.c [hunk 4]
```diff
@@ -10,7 +10,7 @@
 #define HSIZE 324
 
 
-void read_csv(char filename[], int rows, int cols, int array[rows*cols]) {
+void read_csv(char filename[], int rows, int cols, char array[rows*cols]) {
```
Comment (jwg4) [inline tests/test_legacy_dlx.c:13]: l19: the cast to int should be changed to cast to char?

l23: int -> char, avoiding an implicit cast?

**Review by jwg4 — CHANGES_REQUESTED** [review]

**Review by jwg4 — CHANGES_REQUESTED** [review]
Thank you so much for this.

**Comment by jwg4** [comment]
I really appreciate you doing this work.

I have released the code on this branch to pypi as version 1.1.0a0 (a prerelease version). If you like, you can change the version number in pyproject.toml to 1.1.0, and when this branch is merged it will be released as that version. You don't have to do this as I can update the version myself when it's on master.

**Review by jwg4 — APPROVED** [review]

File: exact_cover/wrapper.py (hunk 3 above)
Comment (johnrudge) [reply exact_cover/wrapper.py:8]: Test added!

File: src/exact_cover.c [hunk 5]
```diff
@@ -39,13 +39,13 @@ static PyObject* get_exact_cover(PyObject* self, PyObject* args)
     /*  Parse single numpy array argument */
```
Comment (jwg4) [inline src/exact_cover.c:39]: just seen l37:
`int *in_array_data, rows, cols, result;`
which also needs to be fixed, sorry!

File: tests/test_legacy_dlx.c (hunk 4 above)
Comment (johnrudge) [reply tests/test_legacy_dlx.c:13]: Changes made!

File: src/exact_cover.c (hunk 5 above)
Comment (johnrudge) [reply src/exact_cover.c:40]: Fixed that too!

**Comment by johnrudge** [comment]
I think I'm done tweaking. I extended the test array to cover python versions up to 3.10. I did try 3.11 too but ran into an issue with numpy versions, so that's one to look at another time. Thanks for maintaining the exact-cover package!

**Pull request merged**